### DISCUSSION TITLE
Feature Request:  M codes required for OpenPnP

### DISCUSSION THREAD

**Discussion by phomann** [terjeio/grblHAL#222] [Ideas]
Hi all,
I'm in the process of converting a Yamaha YM84 pick and place machine over to OpenPnP. The original controller from the 80s died and it's getting too hard to repair. The machine has a conventional X, Y axis with rapids of 60m/min. It also has a rotational axis that rotates the nozzles on the three heads. One of the heads also has a Z axis, the other two are pneumatic, just fully up or down.

Additionally, the machine has about 40 pneumatic solenoids or relay outputs plus about the same in inputs. I plan to use a second I/O processor to deal with a lot of these.

I've been looking around at the various G-Code interpreter/controllers such as grbl, Marlin, RepRap, etc but they are too 3D printer specific. GRBL seems the obvious choice and grblHAL along with the Teensy4.1 is the way I'd like to proceed with. Currently I'm using Phil's (thanks Phil) breakout board as the testbed to do my testing and prototyping. Eventually I'll roll my own custom board for the Teensy4.1 that will suit my I/O needs and setup.

OpenPnP relies on several M codes that grblHAL is not currently supporting. These are.

M114 Get Current Position       https://marlinfw.org/docs/gcode/M114.html
M115 Firmware Info                  https://marlinfw.org/docs/gcode/M115.html
M204 Set Starting Acceleration https://marlinfw.org/docs/gcode/M204.html
M400 Finish Moves                    https://marlinfw.org/docs/gcode/M400.html

I had a bit of an attempt to "hack" adding in the M114 and M115 codes without really knowing what I was doing. My guess is that the M115 would be ok as I added in a new report. I did the same for M14 but I'm not sure if I was doing it correctly as I was just calling the report function I wrote (in report.c) directy from the G-code parser function. I'm not sure if that interferes with grblHAL internals.

As to M204 and M400 I don't have anywhere enough understanding of grbl internals to start.
Presumably, these should be added in as a openPnP plugin?

My feature request is to see if I can get these M codes added in. I'm happy to give it a go if I can get enough guidance to assist me through the process. I've done embedded programming in the past, but I didn't what to turn this into a major project for me as I'm trying to get my pick and place machine operational as soon as possible.

Cheers,

Peter

**Comment by terjeio** [comment]
I am not sure about M204 but I believe the others can be implemented in a plugin without changing anything in the core.
The [test branch](https://github.com/terjeio/grblHAL/tree/test) has several changes to internals for supporting plugin (and driver) extensions, do _not_ use the current master as a base when adding any.

I do not know if you have taken a look at existing [plugins](https://github.com/terjeio/grblHAL/tree/test/plugins) and [template](https://github.com/terjeio/grblHAL/tree/test/templates) code. If not I urge you to do so.

The laser [coolant plugin](https://github.com/terjeio/grblHAL/blob/test/plugins/laser/coolant.c) is perhaps the best example.

M400 is doing the same as G4P0? It can be implemented by calling `protocol_buffer_synchronize()` on a M400.

I can provide further guidance for writing a plugin.

**Comment by phomann** [comment]
Thanks for the reply, I'll have a go at the M114 and M115 first. I'll also test the G4P0. It makes sense that this could be the solution for the wait.
I need to sort out a servo H/W issue first before I start. :(
Cheers,
Peter

**Comment by terjeio** [comment]
Here is a skeleton plugin for M114, M115 and M400 - M400 output needs to be added when the format is found out.

[openpnp.zip](https://github.com/terjeio/grblHAL/files/6425356/openpnp.zip)

Drop the openpnp.c file in the same folder as _driver.c_ - no other changes requred as it implements `my_plugin_init()`.

**Reply by terjeio** [reply]
I've done a bit of research and modified the code above from what I found and added it [as a plugin](https://github.com/grblHAL/Plugin_OpenPNP) in the new repo. It is not yet complete, I need someone commited to testing it to do that.

**Comment by phomann** [comment]
Thanks very much for adding that. It's very much appreciated. I'll let you know how I go, although I'm still sorting out a few hardware issues.
Cheers,
Peter

**Reply by terjeio** [reply]
Ok. I have already added code for modifying acceleration via M204 to the core, to be commited later.

M105 for reading analog sensors is possible, but hard to find usable documentation. Is gain/offset settings required for each sensor? Or is this handled by OpenPNP?

**Reply by phomann** [reply]
OpenPnP requires the user to set up a regular expression to define what the format of the returned value is. OpenPnP uses the returned value and compares it to a user defined upper and lower limit to see if it is in an acceptable range. Also, it can handle real or integer values. Any gain and offset can be handled by OpenPnP
So, it may be best to either return the raw A/D integer value, 0-255 for 8 bit, 0-1023 for 12 bit, etc or a real value 0.0 - 100.00 percent value if we want to abstract the hardware resolution from the returned value.
I'm not sure which is best. :)

For my YM84 Yamaha machine, it has a pressure sensor board which converts the pressure to a 2-bit digital value for off, Lo, Med, Hi pressure. For three heads, that's 6 I/O lines. :(  OpenPnP allows for multiple (Sub) controllers. My plan is to have a sub controller that is responsible for most of the pneumatic actuators and sensors. For the vacuum sensors, it will read the 2-bit value and convert it to an analogue value that will be returned to mimic the analog reading of a sensor.
On that note, does anyone know of a suitable G-Code interpreter that I can use to hack as I don't need all the motion stuff, just the Reading and setting of I/O ports, analog ports etc
Cheers,
Peter

**Reply by terjeio** [reply]
> Any gain and offset can be handled by OpenPnP

Ok, that makes it simpler.

> So, it may be best to either return the raw A/D integer value, 0-255 for 8 bit, 0-1023 for 12 bit, etc or a real value 0.0 - 100.00 percent value if we want to abstract the hardware resolution from the returned value.
I'm not sure which is best. :)

Integer value is best as the [ioports API](https://github.com/grblHAL/Templates/blob/master/ioports.c) is based on LinuxCNC specs which is integer only. Any conversion has to be added in a plugin, either hardcoded or via custom $-settings. BTW plugins can add custom settings as they see fit.

> My plan is to have a sub controller that is responsible for most of the pneumatic actuators and sensors.

This could be as simple as a few I2C devices? There is a I2C port on Phils board and the coming Pro version has space for a I2C based daughter board. A suitable map file and corresponding code to expose the ports via the standard ioports API is needed. If done like this the OpenPNP plugin will be able to interact with the ports with no code changes.

> On that note, does anyone know of a suitable G-Code interpreter that I can use to hack as I don't need all the motion stuff, just the Reading and setting of I/O ports, analog ports etc

The grblHAL core can be used together with a custom driver that only supports the ioports API...

**Reply by terjeio** [reply]
> This could be as simple as a few I2C devices?

I forgot to mention that this could be a microcontroller too, I have done that for my Trinamic CNC BoosterPack. It has a [MSP430 controller](https://github.com/terjeio/Trinamic_TMC2130_I2C_SPI_Bridge) used for the Trinamic SPI interface and talks I2C with the main processor. This can also monitor inputs for fault conditions and interrupt the main processor when anything is amiss.

**Reply by phomann** [reply]
Using integers for analog values will be fine.

The sub controller in OpenPnP is just another controller connecting to OpenPnP via a separate USB or Ethernet port. It doesn't talk to the main controller. My current plan is to use another Teensy4.1, although it would be overkill as I could just use a less powered Arduino.
Using the grblHAL core and configuring it with Just digital IO sounds like the way to go.
